From 4ced7179db168cfe3d79321d3592d54892680f2f Mon Sep 17 00:00:00 2001
From: Adrian DC <radian.dc@gmail.com>
Date: Sat, 27 Aug 2016 15:09:36 +0200
Subject: [PATCH] linker: Make platform text relocations denial optional

 * Use the TARGET_NEEDS_PLATFORM_TEXT_RELOCATIONS := true
    configuration to allow a device to use legacy proprietary
    libraries like camera on non-user build variants

 * Partial revert "Remove textrels support for platform libs"
    commit 8068786ae67835291521e52f39c695e40f3ad20d.

 * Forward-port to AOSP 8.0, use __ANDROID_API_M__
    and adapt to Android.bp build rules

Change-Id: I994ab1a600a0b237b496ceebe2dd54febc28a6bd
Signed-off-by: Adrian DC <radian.dc@gmail.com>
---
 linker/linker.cpp | 15 ---------------
 1 file changed, 15 deletions(-)

diff --git a/linker/linker.cpp b/linker/linker.cpp
index 9094b0a4c..c1acf0bd2 100644
--- a/linker/linker.cpp
+++ b/linker/linker.cpp
@@ -3553,21 +3553,6 @@ bool soinfo::link_image(const soinfo_list_t& global_group, const soinfo_list_t&
 
 #if !defined(__LP64__)
   if (has_text_relocations) {
-    // Fail if app is targeting M or above.
-    int app_target_api_level = get_application_target_sdk_version();
-    if (app_target_api_level >= __ANDROID_API_M__) {
-      DL_ERR_AND_LOG("\"%s\" has text relocations (https://android.googlesource.com/platform/"
-                     "bionic/+/master/android-changes-for-ndk-developers.md#Text-Relocations-"
-                     "Enforced-for-API-level-23)", get_realpath());
-      return false;
-    }
-    // Make segments writable to allow text relocations to work properly. We will later call
-    // phdr_table_protect_segments() after all of them are applied.
-    DL_WARN_documented_change(__ANDROID_API_M__,
-                              "Text-Relocations-Enforced-for-API-level-23",
-                              "\"%s\" has text relocations",
-                              get_realpath());
-    add_dlwarning(get_realpath(), "text relocations");
     if (phdr_table_unprotect_segments(phdr, phnum, load_bias) < 0) {
       DL_ERR("can't unprotect loadable segments for \"%s\": %s", get_realpath(), strerror(errno));
       return false;
-- 
2.17.0

